keras/layers/layer.py

```diff
@@ -808,6 +810,16 @@ def maybe_convert(x):
                         outputs = super().__call__(*args, **kwargs)
                 else:
                     outputs = super().__call__(*args, **kwargs)
+                # Change the layout for the layer output if needed.
+                # This is useful for relayout intermediate tensor in the model
+                # to achieve the optimal performance.
+                distribution = distribution_lib.distribution()
+                if distribution is not None:
+                    current_layer_path = current_path()
+                    layout = distribution.get_tensor_layout(current_layer_path)
```

Comment:
So for setting the layout of the output of a `Dense` layer in a subclassed model, you'd do like `layout_map["model/layers/dense_1"] = (...)`?

1. Is there a risk of confusing variable layouts and intermediate tensor layouts?
2. Should we be more specific, e.g. `layout_map["model/layers/dense_1/output"] = (...)` ? This could also leave the door open for `input` if ever needed.

Reply:
> So for setting the layout of the output of a `Dense` layer in a subclassed model, you'd do like `layout_map["model/layers/dense_1"] = (...)`?
> 
> 1. Is there a risk of confusing variable layouts and intermediate tensor layouts?
It does. And my original intent was actually same as option 2.

> 2. Should we be more specific, e.g. `layout_map["model/layers/dense_1/output"] = (...)` ? This could also leave the door open for `input` if ever needed.
That will definitely make the it more explicit. And also, this open the option for mapping to any intermediate keras operations within the layer body.

> 3. Is the full path too much information? What about `layout_map["dense_1/output"] = (...)`? Is that confusing?
Matt has the same question, and he was proposing the use regex.search instead of regex.match so that user can skip the prefix. My original implementation was trying to be a bit strict, so that the layout won't accidentally map to unwanted weights. In the case that there are overlapping rule that apply to same weights, currently the first one wins. Maybe we can take the regex.match approach, and raise an error when multiple rules is mapped to the same weights/tensor. (Probably I will do this in a separate PR.)

Reply:
> Maybe we can take the regex.match approach, and raise an error when multiple rules is mapped to the same weights/tensor

I think that's a good idea, we can use `search` and then make sure that each variable matches at most 1 rule.

Reply:
Ack, let me do this in a separate PR.

Reply:
@mattdangerw .